ouroboros-consensus/docs/test-consensus/Plan.md

```diff
@@ -0,0 +1,629 @@
+because the chains selected by the implementation's nodes as of the onset of a
+slot are not analogous to those selected by the paper's parties. This is
+because nodes do not exactly implement the π_SPoS protocol.
+
+The key difference is that -- message delays permitting -- nodes can select a
+block in the same slot it is received, whereas parties must wait until the
+subsequent slot. Whereas the paper's protocol specifies that a party selects
+from among its received chains once per slot, the implementation's node
+re-selects its chain immediately upon receiving a new block. The simplest
+example is a two-node network with the first slot being led by exactly one
+node. The leader will forge its block once the slot starts and immediately
+select that block. The other node will receive the block almost immediately
+(assuming no network latencies), and then immediately select that same 1-block
+chain, since it's longer than the genesis chain. Therefore both nodes have
+selected the new chain as of the onset of the second slot. The analogous
+example in the paper also has the leader forge a block, immediately select it,
+and send a message to the other node. As we have interpreted the paper, The
+Diffuse Functionality will not deliver that block until the end of the current
+slot, by which time the node will have already checked its inbox for this slot,
+and so it won't consider the new chain for selection until the second slot.
+Only the first slot's leader will have selected the new chain as of the onset
+of second slot. This discrepancy is depicted as follows.
+
+```
+onset of slot:            0    1
+
+In implementation
+leader's selection:       g    g,B
+recipient's selection:    g    g,B
+
+In paper
+leader's selection:       g    g,B
+recipient's selection:    g    g
+```
+
+The only way the π_SPoS protocol could permit the recipient to select the chain
+in the same slot it was forged is if the recipient was asleep/blocked/etc until
+after the Diffuse Functionality had delivered the message. In our judgment, the
+paper does not consider such computational delays; otherwise, the π_SPoS
+protocol as written would also permit a leader to forge a block whose
+predecessor was forged in the same slot. That would violate Definition 6's
+_strictly increasing labels_ invariant, and the π_SPoS protocol does not seem
+to guard against that in any way other than by the assumption that the
+recipient will have checked its inbox before the requisite message arrives. In
+other words, the paper assumes that parties are perfectly synchronized (no
+clock skew) and execute the π_SPoS protocol in total isolation _during_ each
+slot, while messages are instead only delivered _between_ slots.
+
+(Remark. With computational delays and/or clock skews, the current
+implementation's node could forge a block with the same slot as its
+predecessor. However, the current test suite has no clock skews etc, and the
```

Comment:
We don't: https://github.com/input-output-hk/ouroboros-network/blob/9d825e8f541934aa0efffa5039620202c8a4ab0e/ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs#L337